src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1318 @@
+         server_monitor, &ismaster_reply, &start_us, &error);
+      bson_destroy (&ismaster_reply);
+   }
+
+   if (server_monitor->stream) {
+      MONITOR_LOG (server_monitor, "rtt polling ismaster");
+      ret = _server_monitor_polling_ismaster (
+         server_monitor, &ismaster_reply, &error);
+      if (ret) {
+         *rtt_ms = (_now_us () - start_us) / 1000;
+      }
+      bson_destroy (&ismaster_reply);
+   }
+   return ret;
+}
+
+/* The RTT monitor thread function.
+ *
+ * Server monitor must be in state MONGOC_THREAD_OFF.
+ */
+static BSON_THREAD_FUN (_server_monitor_rtt_thread, server_monitor_void)
+{
+   mongoc_server_monitor_t *server_monitor;
+
+   server_monitor = (mongoc_server_monitor_t *) server_monitor_void;
+
+   while (true) {
+      int64_t rtt_ms;
+      bson_error_t error;
+
+      bson_mutex_lock (&server_monitor->shared.mutex);
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         bson_mutex_unlock (&server_monitor->shared.mutex);
+         break;
+      }
+      bson_mutex_unlock (&server_monitor->shared.mutex);
+
+      _server_monitor_ping_server (server_monitor, &rtt_ms);
+      if (rtt_ms != -1) {
+         mongoc_server_description_t *sd;
+
+         bson_mutex_lock (&server_monitor->topology->mutex);
+         sd = mongoc_topology_description_server_by_id (
+            &server_monitor->topology->description,
+            server_monitor->description->id,
+            &error);
+         if (sd) {
+            /* If the server description has been removed, the RTT thread will
+             * be terminated by background monitoring soon. */
+            mongoc_server_description_update_rtt (sd, rtt_ms);
+         }
+         bson_mutex_unlock (&server_monitor->topology->mutex);
+      }
+      mongoc_server_monitor_wait (server_monitor);
+   }
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.state = MONGOC_THREAD_JOINABLE;
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+   BSON_THREAD_RETURN;
+}
+
+void
+mongoc_server_monitor_run (mongoc_server_monitor_t *server_monitor)
+{
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   if (server_monitor->shared.state == MONGOC_THREAD_OFF) {
+      server_monitor->is_rtt = false;
+      server_monitor->shared.state = MONGOC_THREAD_RUNNING;
+      COMMON_PREFIX (thread_create)
+      (&server_monitor->thread, _server_monitor_thread, server_monitor);
+   }
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+void
+mongoc_server_monitor_run_as_rtt (mongoc_server_monitor_t *server_monitor)
+{
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   if (server_monitor->shared.state == MONGOC_THREAD_OFF) {
+      server_monitor->is_rtt = true;
+      server_monitor->shared.state = MONGOC_THREAD_RUNNING;
+      COMMON_PREFIX (thread_create)
+      (&server_monitor->thread, _server_monitor_rtt_thread, server_monitor);
+   }
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
```

Comment:
I reused a lot of the logic for the RTT thread, which differs from the pseudocode proposed in https://github.com/mongodb/specifications/pull/807

There's a discussion between Shane and I here explaining why this diverges: https://github.com/mongodb/specifications/pull/807#discussion_r427647840

The pseudocode assumes ServerDescriptions are immutable, and that they are constructed with the running average RTT. In libmongoc, server descriptions hold the running average, and it is updated in place.